improver/calibration/quantile_regression_random_forest.py

```diff
@@ -0,0 +1,667 @@
+                c.name()
+                for c in template_cube.coords(
+                    dimensions=template_cube.coord_dims(coord)
+                )
+            ]
+            feature_associated_coords = list(
+                set(associated_coords).intersection(features)
+            )
+            if feature_associated_coords:
+                refined_dim_coord_names.append(feature_associated_coords[0])
+            elif coord.name() in dim_coord_names:
+                refined_dim_coord_names.append(coord.name())
+    refined_dim_coord_names = list(set(refined_dim_coord_names))
+
+    expansion_dim_names = list(set(refined_dim_coord_names) - set(features))
+
+    # If the expansion_dims and refined_dim_coord_names match, this implies that
+    # the features provided are not dimension coordinates nor associated with any
+    # dimension coordinate. In this case, any singleton dimensions on the
+    # expansion_array should be removed, as it isn't relevant for defining the shape
+    # of the template cube.
+    if expansion_dim_names == refined_dim_coord_names:
+        expansion_array = np.squeeze(expansion_array)
+
+    dims = []
+    for dim_name in expansion_dim_names:
+        dims.extend(template_cube.coord_dims(dim_name))
+
+    return np.expand_dims(expansion_array, dims)
+
+
+def prep_feature(
+    template_cube: Cube,
+    feature_cube: Cube,
+    feature: str,
+) -> np.ndarray:
+    """Prepare the feature values for the quantile regression random forest model.
+    Each expected feature is handled separately. The key steps are:
+        - Collapse the template cube over the realization dimension to get a shape
+        template.
+        - Extract the feature values from the feature cube provided.
+        - Handle different types of features separately (mean, std, spatial coordinates,
+        forecast period, model weights, day of year, hour of day, and static features)
+        to ensure they are broadcasted correctly to match the template cube shape.
+        Time coordinates are the most complex to handle as they can be scalar, span a
+        single dimension or span multiple dimensions.
+        - Flatten the feature values to create a 1D array.
+        - Set the output dtype based on the feature type.
+        - Return the flattened array of feature values.
+
+    Args:
+        template_cube (cube):
+            The forecast cube that acts only as a template.
+        feature_cube (cube):
+            Feature cube that is either static or dynamic.
+        feature:
+            The feature to be extracted from the associated feature_cube.
+            If "static" then the cube itself acts as the feature.
+    Returns:
+        feature_values (numpy.ndarray):
+            Flattened array of feature values.
+    """
+    # Collapse the template cube to get a shape template.
+    collapsed_cube = template_cube.collapsed(["realization"], iris.analysis.MEAN)
+
+    # Handle each expected feature type separately.
```

Comment:
As discussed separately it feels as though a huge amount of the cube dimension handling below could be done away with by working directly with parquet files rather than translating these to netcdf.